### PR TITLE
[TT-TO-TTGWARP] Detect and handle flash attention with causal masking

### PR DESCRIPTION
**Pull request by jopperm**
Detect and handle flash attention with causal masking in `-convert-triton-to-tritongpu-warp` by supporting `tt.make_range` and *two* dependent attention-`for`-loops.

See #1947 for more context / the complete PoC.

### CODE REVIEW

File: test/Conversion/intel/triton_to_tritongpu_warp.mlir [hunk 1]
```diff
@@ -184,3 +188,153 @@ module {
...
+// -----
+
+// COM: FlashAttention with causal masking
+// COM: - two loops with workload=attention (=4) are detected
+// COM: - encodings are propagated to the operations that correspond to the causal mask computation
+
+// CHECK1: #triton_gpu.blocked<{sizePerThread = [16, 64], threadsPerWarp = [1, 1], warpsPerCTA = [8, 1], order = [1, 0]}>
+// CHECK1: "triton_gpu.num-warps" = 8
+module {
+  tt.func public @_attn_fwd(%arg0: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg3: f32, %arg4: !tt.ptr<f32> {tt.divisibility = 16 : i32}, %arg5: !tt.ptr<f32> {tt.divisibility = 16 : i32}) attributes {noinline = false} {
```
Comment (etiotto) [inline test/Conversion/intel/triton_to_tritongpu_warp.mlir:201]: Remove unnecessary attributes (like the `noinline` one).

File: third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp [hunk 2]
```diff
@@ -383,6 +385,51 @@ class ConvertTritonToTritonGPUWarp
         }
         return WalkResult::advance();
       });
+
+      if (loops.size() == 2 && workloads.front() == Workload::Attention &&
```
Comment (etiotto) [inline third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp:389]: why the loop size has to be 2 ?

File: third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp [hunk 3]
```diff
@@ -383,6 +385,51 @@ class ConvertTritonToTritonGPUWarp
...
+
+          if (auto cst = dyn_cast<arith::ConstantOp>(op)) {
+            transformArithConstantOp(cst, blockLayout);
+            return;
+          }
+
+          // Assign:
+          // - rank-2 operations: block layout
+          // - rank-1 operations: slice layout
+          assert(op->getNumResults() == 1);
```
Comment (etiotto) [inline third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp:416]: Add a msg to the assert stmt

File: third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp [hunk 4]
```diff
@@ -414,9 +461,12 @@ class ConvertTritonToTritonGPUWarp
...
       auto src = expand.getSrc();
-      auto attr = cast<ttg::SliceEncodingAttr>(src.getType().getEncoding());
-      Type newType = addAttrToType(result.getType(), attr.getParent());
-      result.setType(newType);
+      if (auto attr = dyn_cast_if_present<ttg::SliceEncodingAttr>(
+              src.getType().getEncoding())) {
+        Type newType = addAttrToType(result.getType(), attr.getParent());
+        result.setType(newType);
+      }
+      // else: will patch up later
```
Comment (etiotto) [inline third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp:469]: Add a FIXME so it is easy to "grep"

File: test/Conversion/intel/triton_to_tritongpu_warp.mlir [hunk 5]
```diff
@@ -1,5 +1,9 @@
-// RUN: triton-opt %s -split-input-file --convert-triton-to-tritongpu-warp="num-warps=32"  | FileCheck %s --check-prefix=CHECK
-// RUN: triton-opt %s -split-input-file --convert-triton-to-tritongpu-warp="num-warps=8"  | FileCheck %s --check-prefix=CHECK1
+// RUN: triton-opt %s -split-input-file --convert-triton-to-tritongpu-warp="num-warps=32"  | FileCheck %s --check-prefix=CHECK --implicit-check-not="tensor<{{[0-9]+x[0-9]+x[if][0-9]+}}>"
+// RUN: triton-opt %s -split-input-file --convert-triton-to-tritongpu-warp="num-warps=8"  | FileCheck %s --check-prefix=CHECK1 --implicit-check-not="tensor<{{[0-9]+x[0-9]+x[if][0-9]+}}>"
+
+// COM: The implicit-check-not ensures that an encoding attribute is added to all 2D tensors.
```
Comment (Dewei-Wang-sh) [inline test/Conversion/intel/triton_to_tritongpu_warp.mlir:4]: not used before, amazing.

File: test/Conversion/intel/triton_to_tritongpu_warp.mlir [hunk 6]
```diff
@@ -184,3 +188,153 @@ module {
...
+      %54 = tt.expand_dims %51 {axis = 1 : i32} : tensor<128xf32> -> tensor<128x1xf32>
+      %55 = tt.broadcast %54 : tensor<128x1xf32> -> tensor<128x64xf32>
+      %56 = arith.mulf %arg8, %55 : tensor<128x64xf32>
+      %57 = tt.load %arg10 : !tt.ptr<tensor<64x64xf16>>
+      %58 = arith.truncf %48 : tensor<128x64xf32> to tensor<128x64xf16>
+      %59 = tt.dot %58, %57, %56, inputPrecision = tf32 : tensor<128x64xf16> * tensor<64x64xf16> -> tensor<128x64xf32>
+      %60 = tt.advance %arg10, [%c64_i32, %c0_i32] : <tensor<64x64xf16>>
+      %61 = tt.advance %arg11, [%c0_i32, %c64_i32] : <tensor<64x64xf16>>
+      scf.yield %53, %59, %43, %60, %61 : tensor<128xf32>, tensor<128x64xf32>, tensor<128xf32>, !tt.ptr<tensor<64x64xf16>>, !tt.ptr<tensor<64x64xf16>>
+      // CHECK1: workload = 4
```
Comment (Dewei-Wang-sh) [inline test/Conversion/intel/triton_to_tritongpu_warp.mlir:276]: we may change the magic number later.

File: third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp (hunk 2 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp:389]: flashAttention with causal mask has 2 loops.

File: third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp [hunk 7]
```diff
@@ -383,6 +385,51 @@ class ConvertTritonToTritonGPUWarp
...
+            if (auto tty = dyn_cast<RankedTensorType>(ty))
+              if (!tty.getEncoding())
+                typesWithoutEncoding.push_back(tty);
+          }
+
+          if (typesWithoutEncoding.empty())
+            return;
+
+          if (auto cst = dyn_cast<arith::ConstantOp>(op)) {
+            transformArithConstantOp(cst, blockLayout);
```
Comment (Dewei-Wang-sh) [inline third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp:409]: here just assume the type w/o encoding has blockLayout, right?
I have a local change that aims to cover the full propagation, making most of the case go to L405 early return.

**Review by Dewei-Wang-sh — APPROVED** [review]
overall LGTM

File: test/Conversion/intel/triton_to_tritongpu_warp.mlir (hunk 1 above)
Comment (jopperm) [reply test/Conversion/intel/triton_to_tritongpu_warp.mlir]: Done (also in the existing test cases in the same file).

File: third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp (hunk 7 above)
Comment (jopperm) [reply third_party/intel/lib/TritonToTritonGPUWarp/TritonToTritonGPUWarpPass.cpp:412]: Yes, this IR walk just patches previously unhandled ops, but is completely specific to causal flash attention. I added a `FIXME` above to make that clearer. So great if we could drop this workaround soon.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**